MSC2753: Peeking via sync (take 2) (#2753)
matrix-org/matrix-doc: proposals/2753-peeking-via-sync-v2.md

nicely, the act of calling `/sync` without a `since` token must cancel any peeks
created by that device.

Similar to `/join`, `/peek` lets you specify `server_name` querystring
parameters to specify which server(s) to try to peek into the room via (when
coupled with [MSC2444](https://github.com/matrix-org/matrix-doc/pull/2444)).

If a user subsequently `/join`s the room they're peeking, we atomically move
the room to the `join` block of the `/sync` response, allowing the client to
build on the state and history it has already received without re-sending it
down `/sync`.

To stop peeking, the user calls `/unpeek` on the room, similar to `/leave` or
`/forget`.  This returns 200 on success, 404 on unrecognised ID, or 400 if the
room was not being peeked in the first place.  Having stopped peeking, the
empty room block appears in the `leave` block of the next sync response to tell
the client that the user is no longer peeking.

The new `/peek` and `/unpeek` endpoints require authentication and can be
ratelimited. Their responses are analogous to their `/join` and `/leave`
counterparts (eg: `room_id` in the response and empty object when stopping).

> The act of joining a peeked room automatically cancels any ongoing `/peeks` by

Reviewer: this is a duplicate of the above: will remove.

# MSC2753: Peeking via Sync (Take 2)

## Problem

Peeking into rooms without joining them currently relies on the deprecated v1 /initialSync and /events APIs.

This poses the following issues:

 * Servers and clients must implement two separate sets of event-syncing logic, doubling complexity.
 * Peeking a room involves setting a stream of long-lived /events requests
   going. Having multiple streams is racey, competes for resources with the
   /sync stream, and doesn't scale given each room requires a new /events
   stream.
 * v1 APIs are deprecated and not implemented on new servers.

This MSC likely obsoletes [MSC1776](https://github.com/matrix-org/matrix-doc/pull/1776).

> ## Proposal

Reviewer 2: It's not spelled out, but presumably this proposal relies on the access token's associated device to function: if Alice uses token 1 to peek, token 2's sync stream won't have the peek - just token 1's.

Author: It tries to spell it out on lines 24-25?

> "Adds the room (if permissions allow) to a new section of the /sync response called `peek` - but only for the device which called `/peek`.

Reviewer: seems pretty clear to me.

parameters to specify which server(s) to try to peek into the room via (when
coupled with [MSC2444](https://github.com/matrix-org/matrix-doc/pull/2444)).

If a user subsequently `/join`s the room they're peeking, we atomically move
the room to the `join` block of the `/sync` response, allowing the client to
build on the state and history it has already received without re-sending it
down `/sync`.

To stop peeking, the user calls `/unpeek` on the room, similar to `/leave`.
This returns 200 on success, 404 on unrecognised ID, or 400 if the room was not
being peeked in the first place.  Having stopped peeking, the empty room block
appears in the `leave` block of the next sync response to tell the client that
the user is no longer peeking.

The new `/peek` and `/unpeek` endpoints require authentication and can be
ratelimited. Their responses are analogous to their `/join` and `/leave`
counterparts (eg: `room_id` in the response and empty object when stopping).

Clients should check for any irrelevant peeked rooms on launch (left over from
previous instances of the app) and explicitly `/unpeek` them to conserve
resources.

> ## Potential issues

Reviewer 2: It's not clear what the sync section looks like. Is it like `join`? does that mean members can be lazy-loaded?

Author: as line 26 says:

>  * The `peek` section is identical to `join`, but shows the live activity of rooms for which that device is peeking.

so yes, LL would be supported.

    data.
 3. [MSC1772](https://github.com/matrix-org/matrix-doc/pull/1772) (Groups as
    rooms) uses room joins to indicate your own membership, and peeks to query
    the group membership of other users.  Similarly to profiles, it's not clear
    that this should be per-user rather than per-device (and worse case, it's a
    matter of effectively opting in rather than trying to filter out peeks you
    don't care about).

## Alternatives

[MSC1776](https://github.com/matrix-org/matrix-doc/pull/1776) defined an
alternative approach, where you could use filters to add peeked rooms into a
given `/sync` response as needed.  This however had some issues:

 * You can't specify what servers to peek remote rooms via.
 * You can't identify rooms via alias, only ID
 * It feels hacky to bodge peeked rooms into the `join` block of a given
   `/sync` response
 * Fiddling around with custom filters feels clunky relative to just calling a
   `/peek` endpoint similar to `/join`.

While experimenting with implementing MSC1776, I came up with this as an
alternative that empirically feels much simpler and tidier.

## Security considerations

> Servers should ratelimit calls to `/peek` to stop someone DoSing the server.

Reviewer 2: Rate limiting unpeek would be good too: 
```suggestion
Servers should ratelimit calls to the new endpoints to stop someone DoSing the server.
```

Author: why would you care about rate limiting unpeek more than any other CS API?  it doesn't add load to the server, it reduces it.

Reviewer 2: it's rate limited up above, so this is more or less reinforcing it. It's also useful to allow servers to rate limit things for repeated requests, even if it doesn't add additional load.

# MSC2753: Peeking via Sync (Take 2)

## Problem

Peeking into rooms without joining them currently relies on the deprecated v1 /initialSync and /events APIs.

This poses the following issues:

 * Servers and clients must implement two separate sets of event-syncing logic, doubling complexity.
 * Peeking a room involves setting a stream of long-lived /events requests
   going. Having multiple streams is racey, competes for resources with the
   /sync stream, and doesn't scale given each room requires a new /events
   stream.
 * v1 APIs are deprecated and not implemented on new servers.

This MSC likely obsoletes [MSC1776](https://github.com/matrix-org/matrix-doc/pull/1776).

> 

Reviewer 3: Please spell out some use cases (and counter-examples).

Matthew listed some briefly in this message: https://matrix.to/#/!TBzpVszyznlSbXpxnR:kb1rd.net/$Qbvq21i_gOYiwao_7sBEX--Rf9yD-lHMGVm0aWUc-4E?via=kb1rd.net&via=matrix.org&via=foad.me.uk . To be useful, to explore the boundaries and requirements, I feel they need a few lines of detail each.

Reviewer 3: I propose this use case:

- As a lowly end-user, I want to subscribe to a public room in order to just read it. I don't yet want to write anything, nor be listed publicly as a "member", nor show my presence or read-markers or profile or anything like that. I want to be able to read the room on all my devices, and manage my subscribed rooms, in the same way as for joined rooms. My participation may be counted and shared in the room or elsewhere as a statistic (e.g. increment a count of observers), while mainly functionally anonymous or invisible, though need not be truly anonymous or invisible. 

This would be particularly useful if we could support a large number of subscribers to a room in this way, more efficiently than joining it.

Reviewer: summarised some usecases. I think huge essays for each point will get in the way here.

## Proposal

We add an CS API called `/peek/{roomIdOrAlias}`, very similar to `/join/{roomIdOrAlias}`.

Calling `/peek`:

>  * Resolves the given room alias to a room ID, if needed.

Reviewer 3: Make it analogous to `/_matrix/client/r0/rooms/{roomId}/join` instead. An API that magically resolves an alias to a room-id ties this portion of the spec to the aliases functionality. Better to allow alias-resolving to be handled by its own spec that can be upgraded independently, and leave it to SDKs to offer "convenience wrappers" rather than putting them in at this level.

Reviewer: I somewhat agree, but `/_matrix/client/r0/rooms/{roomId}/join` does not allow joining rooms over federation. Arguably that could be changed but we should avoid adding unnecessary hurdles for this feature.

Reviewer: I've tweaked the API shape anyway, which rather reopens this question. Let's take further discussion to https://github.com/matrix-org/matrix-doc/pull/2753#discussion_r524393462.

 * Adds the room (if permissions allow) to a new section of the /sync response

>    called `peek` - but only for the device which called `/peek`.

Reviewer 3: Scratch "only for the device". Peeking should be a variant of joining, with as few differences as possible. If there are use cases for functionality of filtering by device, it should be (a) optional, and (b) consistent with the join/sync API.

Reviewer: there's a whole section spelling this out below.

 * The `peek` section is identical to `join`, but shows the live activity of
   rooms for which that device is peeking.

The API returns 404 on an unrecognised room ID or alias, or 403 if the room
does not allow peeking.  Rooms allow peeking if they have `history_visibility`
of `world_readable`.  N.B. `join_rules` do not affect peekability - it's
possible to have an invite-only room which joe public can still peek into, if
`history_visibility` has been set to `world_readable`.

If the `history_visibility` of a room changes to not be `world_readable`, any
peeks on the room are cancelled.

In order to clear up stale peeks if a client restarts without cleaning up

> nicely, the act of calling `/sync` without a `since` token must cancel any peeks

Author: During impl in dendrite it became clear this was a dubious idea - it's a bit obnoxiously magic-at-a-distance that an initial /sync randomly goes and cancels any peeks in progress from that device.

Reviewer: I'm surprised by this. Intuitively I would say that a client does an intial sync if and only if it has dropped its cache - in which case, it will also have forgotten what it is meant to be peeking. What am I missing?

Reviewer 4: It also means that if you initial sync separately (e.g. you stole an access token from the DB to manually debug something?) then it effects clients, which feels like a potential footgun.

Reviewer: well that's fair, I guess, though we've always said that sharing access tokens between devices is dicey (for example: syncing also"acknowledges" delivery of to-device messages, though admittedly not in an initialsync)

Author: i'm ambivalent on it. i agree it's a pretty crappy way to clear up leaky peeks, given clients rarely initial sync. and i agree with erik's point that it's a bit weird that hitting /sync without a since token would have this particular side-effect. (i think the "acknowledge delivery of to-device messages" is an okay side-effect, although it's still a little magical that someone manually invoking incremental /sync could nuke their e2ee)

Reviewer: have removed this para.

created by that device.

Similar to `/join`, `/peek` lets you specify `server_name` querystring
parameters to specify which server(s) to try to peek into the room via (when
coupled with [MSC2444](https://github.com/matrix-org/matrix-doc/pull/2444)).

If a user subsequently `/join`s the room they're peeking, we atomically move
the room to the `join` block of the `/sync` response, allowing the client to
build on the state and history it has already received without re-sending it
down `/sync`.

To stop peeking, the user calls `/unpeek` on the room, similar to `/leave`.
This returns 200 on success, 404 on unrecognised ID, or 400 if the room was not
being peeked in the first place.  Having stopped peeking, the empty room block
appears in the `leave` block of the next sync response to tell the client that

> the user is no longer peeking.

Reviewer 3: "Empty room block"... Presumably intended to be consistent with join-leave-sync behaviour? That's stated in https://matrix.org/docs/spec/client_server/r0.6.1#get-matrix-client-r0-sync as, "The rooms that the user has left or been banned from, mapped as room ID to room information." I don't see that expanded or given in an example, but anyhow the use of identical/consistent wording here, or a reference to there, would be clearer.

Reviewer: :man_shrugging: combining this thread with https://github.com/matrix-org/matrix-doc/pull/2753#discussion_r524396624

*renew* the peek request before this lifetime expires. The server is free to
pick any lifetime.

### Starting a peek

We add an CS API called `/peek_room`, which starts peeking into a given
room. This is similar to
[`/join/{roomIdOrAlias}`](https://matrix.org/docs/spec/client_server/r0.6.1#post-matrix-client-r0-join-roomidoralias)
but has a slightly different API shape.

For example:

```
POST /_matrix/client/r0/peek_room HTTP/1.1

{
    "room": "<room id or alias>",
    "servers": [
        "server1", "server2"
    ]
}
```

A successful response has the following format:

```
{
    "room_id": "<resolved room id>",
    "peek_expiry_ts": 1605534210000
}
```

The `room` parameter is required and must be a valid room id or alias. The
`servers` parameter is optional and, if present, gives a list of servers to try
to peek through.

XXX: should we limit this API to room IDs, and require clients to do a `GET
/_matrix/client/r0/directory/room/{roomAlias}` request if they have a room
alias? (In which case, `/_matrix/client/r0/room/{room_id}/peek` might be a
better name for it.)  On the one hand: cleaner, simpler API. On the other: more

> requests needed for each operation.

Reviewer: Thoughts on this particularly welcome.

Reviewer 5: require an ID for me. If a client is trying to peek an alias, they will have likely already (or should have) called the directory API to confirm it's existence, IMO.

Author: we want peeks to be as fast as possible, and adding in a roundtrip to first lookup the alias is completely unnecessary and so adds latency for no good reason.  So i reckon we should keep it supporting aliases.

Reviewer: (a) if the peek is over federation, there are going to be multiple requests anyway, and it's going to be slow anyway, and having two API calls gives better feedback to the client as to which part failed. (b) As H-S says, in most cases we'll know the ID anyway (in particular #1772 uses IDs everywhere.

active peek). When a peek first starts, the entire state of the room is
included in the same way as when a room is first joined.

If the client requests lazy-loading via `lazy_load_members`, then redundant
membership events are excluded in the same way as they are for joined rooms.

If a user subsequently `/join`s a room they are peeking, the room will
thenceforth appear in the `join` section instead of `peek`. For devices which
were already peeking into the room, the server should *not* include the entire
room state for the room in the `/sync` response, allowing the client to build
on the state and history it has already received without re-sending it down
`/sync`.

### Stopping a peek

To stop peeking, the client calls `rooms/<id>/unpeek`:

```
POST /_matrix/client/r0/rooms/{room_id}/unpeek HTTP/1.1

{}
```

The body must be a JSON dictionary, but no parameters are specified.

A successful response has an empty body.

If the room is unknown or was not previously being peeked the server returns a
400 error with `M_BAD_STATE`.

Having stopped peeking, the unpeeked room will appear in the `leave` block of
the next sync response to tell the client that the user is no longer

> peeking. XXX: why do we do this? Will it be empty?

Reviewer: @ara4n what was the idea of putting it in the `leave` block? The client can tell we've unpeeked via a 200 response to `/unpeek`.

Author: So you can be told if the server has unpeeked you of its own volition (due to a timeout of some kind)

Reviewer: fair. will expand on this.

> # MSC2753: Peeking via Sync (Take 2)

Author: Something this should also handle is supporting /event and /context into peeked rooms.

Reviewer 6: Shouldn't that be accepted/handled automatically as long as a peek is active?

### Starting a peek

We add an CS API called `/peek`, which starts peeking into a given
room. This is similar to
[`/join/{roomIdOrAlias}`](https://matrix.org/docs/spec/client_server/r0.6.1#post-matrix-client-r0-join-roomidoralias)
but has a slightly different API shape.

For example:

```
POST /_matrix/client/r0/peek/{roomIdOrAlias} HTTP/1.1

{
    "servers": [
        "server1", "server2"
    ]
}
```

A successful response has the following format:

```
{
    "room_id": "<resolved room id>",
    "peek_expiry_ts": 1605534210000
}
```

The `servers` parameter is optional and, if present, gives a list of servers to
try to peek through.

XXX: should we limit this API to room IDs, and require clients to do a `GET
/_matrix/client/r0/directory/room/{roomAlias}` request if they have a room
alias? (In which case, `/_matrix/client/r0/room/{room_id}/peek` might be a
better name for it.)  On the one hand: cleaner, simpler API. On the other: more
requests needed for each operation.

Both `room_id` and `peek_expiry_ts` are required in the
response. `peek_expiry_ts` gives a timestamp (milliseconds since the unix
epoch) when the server will *expire* the peek if the client does not renew it.

> The server ratelimit requests to `/peek` and returns a 429 error with

Reviewer 7: ```suggestion
The server should ratelimit requests to `/peek` and returns a 429 error with
```

`M_LIMIT_EXCEEDED` if the limit is exceeded.

Otherwise, the server first resolves the given room alias to a room ID, if
needed.

If there is already an active peek for the room in question, it is renewed and

> a successful response is returned with the updated `peek_expiry_ts`.

Reviewer 7: ```suggestion
a successful response is returned with an updated `peek_expiry_ts`.
```

"an" seems to imply that the client is providing the expiry which is not the case here.

If the user is already *joined* to the room in question, the server returns a

> 400 error with `M_BAD_STATE`.

Reviewer 7: It would be useful to have a specific error code here. Especially if an alias is accepted as the client may not be able that it is joined to the room already. In particular it seems that at least the resolved room ID should be returned.

If the room in question does not exist, the server returns a 404 error with
`M_NOT_FOUND`.

If the room does not allow peeking (ie, it does not have `history_visibility`
of `world_readable` <sup id="a1">[1](#f1)</sup>), the server returns a 403

> error with `M_FORBIDDEN`.

Reviewer 6: Could a rule be added where a user can peek the room if they have been invited to that room? (If `history_visibility` is `shared` or `invited`)

This would allow "previewing the room before joining it", i.e. "do you want to join this room" with some context.

This would also allow clients to implement "accept/reject DM request", and display DMs in a similar manner as they do in many IM clients today; by adding the DM to the DM list as any other DM.

Such a rule could probably synergize with [MSC2199](https://github.com/matrix-org/matrix-doc/pull/2199) and [MSC2444](https://github.com/matrix-org/matrix-doc/pull/2444), providing users more context about the rooms/DMs they're invited to, and be able to accept/reject with more information at hand.

Otherwise, the server starts a peek for the calling device into the given room,
and returns a 200 response as above.

When a peek first starts, the current state of the room is returned in the
`peek` section of the next `/sync` response.

### Stopping a peek

To stop peeking, the client calls `rooms/<id>/unpeek`:

```
POST /_matrix/client/r0/rooms/{room_id}/unpeek HTTP/1.1

{}
```

The body must be a JSON dictionary, but no parameters are specified.

A successful response has an empty body.

If the room is unknown or was not previously being peeked the server returns a

> 400 error with `M_BAD_STATE`.

Reviewer 7: This leads to an unhappy situation where a delayed request may fail because the peek has expired. IIUC this would be hard for for the client to handle gracefully in any other way then just assuming success. Since this is the case why do we differentiate here? Why not make this return success in any case where the client is not peeking when the response is returned. AKA why not make this idempotent.

### `/sync` response

We add a new `peek` section to the `rooms` field of the `/sync`
response. `peek` has the same shape as `join`.

While a peek into a given room is active, any new events in the room cause that
room to be included in the `peek` section (but only for the device with the
active peek). When a peek first starts, the entire state of the room is
included in the same way as when a room is first joined.

If the client requests lazy-loading via `lazy_load_members`, then redundant
membership events are excluded in the same way as they are for joined rooms.

> If a user subsequently `/join`s a room they are peeking, the room will

Reviewer 8: If someone `/peek`s, they'll see the room in the `peek` section and eventually the `leave` section.
If someone `/peek`s, then eventually joins the room, it will further appear in the `join` section.
Now, what if we have these sequence of events.
- Device A `/peek`s into a room.
- Device A  `/sync`s once and gets the (initial) state of the peeked room.
- Device B `/join`s the room that Device A is peeking.
- Device B leaves the room almost immediately.
- Device A `/sync`s again ... and gets what?

According to this paragraph, it's reasonable to infer that Device A will see the room appear in the `leave` section, since the user has left the room but this client has some state that needs updating.

Does this mean my (Device A) device specific `/peek` has been cancelled by this other random device?
Sure, I could `/peek` again but room state can be expensive to get. I already have fairly fresh state locally that just needs an incremental update.

Reviewer 7: The other question is what is the desired UX. I would expect that if I was peeking a room in the context of considering joining I would expect that joining on another device transitioned the view into the room itself as a member. In that case leaving the room should bring me to whatever the "left room" experience is.

However there are other cases for peeking such as spaces or "X as rooms" (such as "profiles as rooms") where you would not want to be unpeeked based on joining the space. However this likely applies for both this device and another device joining the space.

It seems that the best option is the the room is left in the peek list if it is left, whether on this client or another one. Then this client can decide if it still needs that peek (for example it is peeking the space for membership info) or if it is no longer necessary (for example it was for a room preview, and after joining and leaving the user should now be in a "left room" UX where no new info is required).

thenceforth appear in the `join` section instead of `peek`. For devices which
were already peeking into the room, the server should *not* include the entire
room state for the room in the `/sync` response, allowing the client to build
on the state and history it has already received without re-sending it down
`/sync`.

When a room stops being peeked (either because the client called `/unpeek` or
because the server timed out the peek), the room will be included in the

> `leave` section of the `/sync` response, including any events that occured

Reviewer 8: If I'm peeking into a room, then someone invites me to it, then rescinds the invitation, has my peek been cancelled?
How do I differentiate between my invite being rescinded and my peek ending?

Also, if my peek ended, I have to start again, etc. More on this in the previous comment.

between the previous `/sync` and the the peek ending. If there are no such
events, the room's entry in the `leave` section will be empty.

For example:

```js
{
  "rooms": {
    "join": { /* ... */ },
    "leave": {
      "!unpeeked:example.org": {
        "timeline": {
          "events": [
            { "type": "m.room.message", "content": {"body": "just one more thing"}}
          ]
        }
      },
      "!alsounpeeked:example.com": {}
    }
  }
}
```

## Encrypted rooms

(this para taken from MSC #2444):

It is considered a feature that you cannot peek into encrypted rooms, given
the act of peeking would leak the identity of the peeker to the joined users
in the room (as they'd need to encrypt for the peeker). This also feels
acceptable given there is little point in encrypting something intended to be
world-readable.

## Future extensions

 * "snapshot" API, for a one-time peek operation which returns the current
   state of the room without adding the room to future `/sync` responses. Might

>    be useful for certain usecases (eg, looking at a user's public profile)?

Reviewer 9: and for things like matrix-static which do not want to maintain /sync and actively peek over a large set of rooms.

 * "bulk peek" API, for peeking into many rooms at once. Might be useful for
   flair (which requires peeking into lots of users' profile rooms), though
   realistically that usecase will need server-side support.

 * "cross-device" peeks could be useful for microblogging etc?

## Potential issues

 * Expiring peeks might be hard for clients to manage?

## Alternatives

### Keep /peek closer to /join

Given that peeking has parallels to joining, it might be preferable to keep the

> API closer to `/join`. On the other hand, they probabably aren't actually

Reviewer 8: Typo
```suggestion
API closer to `/join`. On the other hand, they probably aren't actually
```

similar enough to make it worth propagating the oddities of `/join` (in
particular, the use of query-parameters
([matrix-doc#2864](https://github.com/matrix-org/matrix-doc/issues/2864)).

### Filter-based API

[MSC1776](https://github.com/matrix-org/matrix-doc/pull/1776) defined an
alternative approach, where you could use filters to add peeked rooms into a
given `/sync` response as needed.  This however had some issues:

 * You can't specify what servers to peek remote rooms via.
 * You can't identify rooms via alias, only ID
 * It feels hacky to bodge peeked rooms into the `join` block of a given
   `/sync` response
 * Fiddling around with custom filters feels clunky relative to just calling a
   `/peek` endpoint similar to `/join`.

### Use the `join` block

It could be seen as controversial to add another new block to the `/sync`
response.  We could use the existing `join` block, but:

 * it's a misnomer (given the user hasn't joined the rooms)
 * `join` refers to rooms which the *user* is in, rather than that they are
   peeking into using a given *device*

>  * we risk breaking clients who aren't aware of the new style of peeking.

Reviewer 9: surely not given that it would still be per-device and those clients wouldn't be calling the new `/peek` API

    per-device - you would not expect peeked rooms to randomly pop up on other
    devices, or consume their bandwidth.
 2. [MSC1769](https://github.com/matrix-org/matrix-doc/pull/1769) (Profiles as
    rooms) is also per device: if a given client wants to look at the Member
    Info for a given user in a room, it shouldn't pollute the others with that
    data.
 3. [MSC1772](https://github.com/matrix-org/matrix-doc/pull/1772) (Groups as
    rooms) uses room joins to indicate your own membership, and peeks to query
    the group membership of other users.  Similarly to profiles, it's not clear
    that this should be per-user rather than per-device (and worse case, it's a
    matter of effectively opting in rather than trying to filter out peeks you
    don't care about).

### Alternatives to expiring peeks

Having servers expire peek requests could be fiddly, so we considered a number
of alternatives:

 * Allow peeks to stack up without limit and trust that clients will not forget
   about them: after all, it is in clients' best interest not to leak
   resources, to reduce the amount of data to be handled, and it is not obvious

>    that leaking peeks is easier than leaking joins.

Reviewer 7: The client cost to "forgetting" a peek is very low. However the server may need to do some work (if `/sync` is poll-based for all rooms). I think the only client cost is if you are offline for a while and get a huge number of events before you get the chance to unpeek a room you no longer need.

One option would be to return all peeked rooms in the sync response even if they have no events. The upside is that you constantly (or periodically) remind the clients what they are peeking but the downside is some extra transfer and parsing (although this can be made arbitrarily low by only sending a "reminder" for each room occasionally).

Currently, peeking relies on the deprecated v1 `/initialSync` and `/events`
APIs.

This poses the following issues:

 * Servers and clients must implement two separate sets of event-syncing logic,
   doubling complexity.
 * Peeking a room involves setting a stream of long-lived `/events` requests
   going. Having multiple streams is racey, competes for resources with the
   `/sync` stream, and doesn't scale given each room requires a new `/events`
   stream.
 * `/initialSync` and `/events` are deprecated and not implemented on new
   servers.

This proposal suggests a new API in which events in peeked rooms would be
returned over `/sync`.

## Proposal

Peeking into a room remains per-device: if the user has multiple devices, each
of which wants to peek into a given room, then each device must make a separate
request.

To help avoid situations where clients create peek requests and forget about
them, each peek request is given a lifetime by the server. The client must
*renew* the peek request before this lifetime expires. The server is free to
pick any lifetime.

### Starting a peek

> We add an CS API called `/peek_room`, which starts peeking into a given

Reviewer: @ara4n wants us to drop the `_room` suffix. I worry that this is somewhat confusing (it's not obvious to the casual observer what it is that we are peeking), and risks a lot of confusion in the future when we find something else to peek, but he seems to feel strongly.

Reviewer: (done)

room. This is similar to
[`/join/{roomIdOrAlias}`](https://matrix.org/docs/spec/client_server/r0.6.1#post-matrix-client-r0-join-roomidoralias)
but has a slightly different API shape.

For example:

```

> POST /_matrix/client/r0/peek HTTP/1.1

Author: It feels really weird to lose the room id or alias from the URL - we won't be able to see it in logs, and given it's the main object of the method, it feels like it should be in the URL rather than buried as a field in the POST request body.

I'm happy to dump the servers into the post body (rather than being querystring params) if it makes folks feel happier though.

Otherwise this is looking good to me.

Reviewer 9: Agreed, why can't this use the same existing shape as `/_matrix/client/r0/join/{roomIdOrAlias}?server_name=x&server_name=y`

Reviewer: I'll stick the room ID/alias back in the URI. As for the `server_name`, we went round this a few times in a chat room, but I'll try to summarise:
 * I think consistency with `/join` is a bit of a false idol - the semantics are necessarily different, we have to add extra fields to the response, and currently (for better or worse) the body of `/join` ends up in the content of the `join` event (cf https://github.com/matrix-org/synapse/issues/8763), which obviously we have no equivalent for here.
 * Unnecessarily mixing query-params and body params seems very weird, and something that I'd like to fix in `/join` (https://github.com/matrix-org/matrix-doc/issues/2864). I'd rather not expand that problem to another API.

{
    "room": "<room id or alias>",
    "servers": [
        "server1", "server2"
    ]
}
```

A successful response has the following format:

```
{
    "room_id": "<resolved room id>",

>     "peek_expiry_ts": 1605534210000

Author: expecting clients to maintain a clock and be NTP synced feels unnecessary. it's also inconsistent with the `renewal_interval` approach that #2444 takes. can't we just return a lifetime for the peek?

Reviewer 10: (Don't we already expect clients to maintain a clock to keep TLS alive for federation?)

Reviewer: > can't we just return a lifetime for the peek?

yes we can, and I was in somewhat two minds about it. `renewal_interval`? `ttl` for consistency with `/voip/turnServer` ? `lifetime` for consistency with `m.call.invite` ?

Reviewer: >  (Don't we already expect clients to maintain a clock to keep TLS alive?)

I'm not following this. Most client impls have nothing to do with TLS.

Reviewer 10: > I'm not following this. Most client impls have nothing to do with TLS.

I mean that if the client's clock is that far out of sync, that they aren't going to be able to set up a TLS connection to the CS API anyway.

Reviewer: oh, I see. Possibly, yes.

Author: so can we make it a `renewal_interval` for consistency with #2444 (and to avoid people having to be NTP synced?)

Reviewer: well I agree that it may as well be consistent with #2444, but we can change #2444 to match whatever we decide here.

`renewal_interval` implies a fixed interval, whereas the server might want to vary the lifetime depending on what else is going on. I'd prefer `lifetime`.

OTOH I don't feel that strongly and don't want to bikeshed it, so if you are keen on `renewal_interval` we can go with that.

If a user subsequently `/join`s the room they're peeking, we atomically move the room to the `joined` block of the `/sync` response, allowing the client to build on the state and history it has already received without re-sending it down /sync.

To stop peeking, the user calls `/unpeek` on the room, similar to `/leave` or `/forget`.  This returns 200 on success, 404 on unrecognised ID, or 400 if the room was not being peeked in the first place.

## Potential issues

It could be seen as controversial to add another new block to the `/sync` response.  We could use the existing `joined` block, but:

 * it's a misnomer (given the user hasn't joined the rooms)
 * `joined` refers to rooms which the *user* is in, rather than that they are peeking into using a given device
 * we risk breaking clients who aren't aware of the new style of peeking.
 * there's already a precedent for per-device blocks in the sync response (for to-device messages)

It could be seen as controversial to make peeking a per-device rather than per-user feature.  When thinking through use cases for peeking, however:

>  1. Peeking a chatroom before joining it is the common case, and is definitely per-device - you would not expect peeked rooms to randomly pop up on other devices, or consume their bandwidth.

Reviewer 11: I'd even say peeking a chatroom is expected only as long as you have your client open. E.g. if I peek, close my client, and open my client again, I'm not sure I would expect the peek to still be there, unless the app remembers my exact location in the UI.

So, one concern here could be that clients (especially without storage or just not cleaning up peeks properly) don't close the peek, and if you come back either the client: 
 - ignores the peek sync response and uses more bandwidth and server resources than needed
 - all of a sudden pops up a peek you have previously opened.

I'm not sure I have a good solution for this though, apart from passing some kind of peek token as well to `/sync`, but expanding the /sync parameters like that might be controversial as well.

Author: If we're passing a peek token to /sync, we kinda end up being back in the `filter` solution from #1776 again.

Is it that bad to assume that any client which knows how to /peek should also be able to go and clean up any unexpected /peeks it sees on launch?

Reviewer 5: Alternatively, the implementation could limit how many active peeks a device is allowed to have ongoing. This would probably help to "enforce" clients clean up, as well as avoiding someone from spinning up 10000 peeks (slowly, due to RL).

Reviewer 11: We might be peeking into a lot of profile rooms on purpose though ...

Reviewer 11: > Is it that bad to assume that any client which knows how to /peek should also be able to go and clean up any irrelevant /peeks it sees on launch?

Hmm, perhaps it could help clients that don't store their peeks to have an API to query their current peeks?

Reviewer 12: With a flag indicating per-user peeks instead of per-device peeks (as suggested in https://github.com/matrix-org/matrix-doc/pull/2753/files#r480026303) this kinda becomes required, because the client can't keep track of peeked rooms on it's own.

Reviewer 4: I'm also a little concerned how easy it would be for clients to end up peeking into a tonne of rooms without cleaning up. I can see it being very easy to forget in some edge cases to do `/unpeek` (e.g. crashes), and to then just ignore any rooms in the peek section that weren't expected (this is somewhat forced as doing `/unpeek` may race with the `/sync` and so you don't want to error/crash in that case). This isn't handled by the current proposal for cleaning out peeks as the device would still be active.

FWIW, the way that this was implemented in my paginated sync branch on synapse (albeit without federated peeks) was a half way house where you did `POST /sync` and had a separate parameter in the body for the rooms that the client wanted to peek. If they wanted to stop peeking then the client would simply stop including the peeking parameter in sync. This has the advantage that if the client forgets that its peeking for whatever reason, it'll stop sending the parameter, implicitly informing the server that its stopped peeking. This method also doesn't share the downsides of re-using filter, as we'd just have a dedicated `peek` section in the request body.

The major downside with this is that you need to encode a `request/response` style API within `/sync` response to deal with errors when trying to peek; though the current proposal would still likely require some way for the server to inform the client that peeking was no longer possible. One way around this is to use a `/peek` API hit, and *then* include the room ID in the request to `/sync`, though its not clear if that is just unnecessarily complex.

On the flip side, this would interact quite nicely with the paginated sync. Firstly, it'll require more options when doing `/sync` to e.g. deal with scrolling, and so a `POST /sync` form is likely desirable. Secondly. one of the problems with paginated sync is that you might click on a room that you haven't seen down `/sync` yet, and so need to be able to inform the server that you want to get info for that room down sync (doing so as a separate API hit runs the risk of races between that and ongoing `/sync`), which would probably look somewhat similar to this `peek` parameter.

Reviewer: > < @jcgruenhage > With a flag indicating per-user peeks instead of per-device peeks (as suggested in https://github.com/matrix-org/matrix-doc/pull/2753/files#r480026303) this kinda becomes required,

*what* becomes required?

Reviewer: > < @erikjohnston > I'm also a little concerned how easy it would be for clients to end up peeking into a tonne of rooms without cleaning up. 

yes, this would be my main concern here as well. Other ideas for how to solve this might include:

 * Allow each device to have only one "peek" active at a time, which can contain a list of rooms to peek into. Each time the client wants to peek into a new room, it should resend the full list of room ids it is currently interested in.
 * Require that each "peek" gets refreshed periodically, say every hour. A `/sync` request could count as a refresh, meaning that clients only need to refresh on restart.

On erik's suggestion of specifying the room list in the sync body: One particular concern might be that it could take many seconds to start peeking into a room over federation, and we don't really want to block sync until that happens. (Maybe it doesn't matter: we just allow the peek to proceed in the background and put a result in a sync result further down the line? But in that case do we still have problems with races?

Reviewer: >     * Require that each "peek" gets refreshed periodically, say every hour. A `/sync` request could count as a refresh, meaning that clients only need to refresh on restart.

It is pointed out in a [different thread](https://github.com/matrix-org/matrix-doc/pull/2753/files#r481102285) that this doesn't work because clients run on computers that get suspended. Ignore that one.

Reviewer: > On erik's suggestion of specifying the room list in the sync body:

another problem with this: if the list of peeked rooms is long, this could amount to a significant amount of data for every /sync request. Honestly it's bad enough that we have to poll rather than use a websocket etc; adding lots more data to the request seems like a retrograde step.

Author: I still think the simplest approach here it to make it the client's problem to clean up after itself, given only it knows how long it needs the peeks for.  If it's smart enough to do something relatively sophisticated like a /peek, it can also go and /unpeek at the right point (e.g. when the user switches room, or restarts the client).  If it doesn't, then the sync will just keep stacking up with peeked rooms, which will be obvious.  And if the client is never run again, then a) the user should eventually remove the device, b) it's not going to be using any resources on the server, given federated peeks time out if nobody is using them.

Reviewer: It can, but will it?

Author: Another option here could be to have `/peek` return a validity period, and the client has to keep re`/peek`ing within that validity period to keep the peek alive - i.e. the same as we do in the SS API in #2444. This is nice because:

 * Peeks clean up after themselves if you forget to do so manually
 * You don't waste bandwidth re-submitting all your ongoing peeks every time you change where you're peeking
 * Symmetrical with SS API

The downside is you need to track timers on the client to remember to keep refreshing your /peeks, which feels a bit fiddly.

Personally I still prefer relying on the client to clean up after itself nicely (which is also minimal bandwidth).

Reviewer: the MSC now proposes that peeks have an expiry. Hopefully this resolves most concerns in this thread without being too obnoxious for clients.

>  2. [MSC1769](https://github.com/matrix-org/matrix-doc/pull/1769) (Profiles as rooms) is also per device: if a given client wants to look at the Member Info for a given user in a room, it shouldn't pollute the others with that data.

Reviewer 11: This could result in _a lot_ of peeks, more than joined rooms (e.g. 1 for every user visible to your account). Are there any scalability concerns here?

Author: I'm assuming that in general you'd only pull in extensible profiles on demand (e.g. by clicking MemberInfo in eleweb).  The exception might be in calculating displaying flair in #1772 - but the (somewhat fuzzy) proposal there is to calculate it serverside for efficiency and to avoid the client having to frantically peek around everywhere.

Reviewer 2: from a client-side perspective, it'd be better for this use case to have `/peek` return an initial state for the room so things like profiles don't have to be dug out of the sync stream. Thinking of the js-sdk in particular, a client would make the request, get a room ID, wait for a Room object that could take forever to show up (harder to time out), then finally deal with the state. If at least `/peek` blocked for the state, the client could rely on a single endpoint.

A `/peek/snapshot` or similar endpoint could also solve this use case, to avoid having to balance peeks and unpeeks constantly.

also, while writing this it occurred to me that the MSC isn't clear how long `/peek` blocks for - is it like `/join` where it waits until it has the full state or just until it knows a `room_id` and can return that?

Reviewer 12: I don't think clients would only peek the profile room if memberinfo is expanded, depending on the client it would likely want to show (part of) that info in the member list already. Same for group rooms, for flair, the client would already want that stuff immediately, once a member of that group is sending a message, but clients without flair don't need this.

Maybe adding flags to the filter for implicit peeking of such rooms would be a good solution here? These could be lazy-loaded to only be sent in a sync response when they are required through a user sending an event.

Reviewer 4: > from a client-side perspective, it'd be better for this use case to have `/peek` return an initial state for the room so things like profiles don't have to be dug out of the sync stream. Thinking of the js-sdk in particular, a client would make the request, get a room ID, wait for a Room object that could take forever to show up (harder to time out), then finally deal with the state. If at least `/peek` blocked for the state, the client could rely on a single endpoint.

Alas, the issue with such APIs is that you then need to somehow link together the `/peek` and `/sync` responses in a way that isn't racey. Its possible to do, but it makes me scream internally at the potential complexity that would have to happen to make it work

Author: hum, somehow i totally missed this discussion back in Sept - sorry.

Yeah, the ability to get a quick snapshot feels very desirable without having to mess around peeking and then unpeeking a room, when you really just want "give me this filtered state from this room please so i can see its profile/name/topic/moderation-events" etc.  It might also be nice for matrix.to.

I guess one advantage of doing peeking by putting filters on /sync would be that you could then do a once-off /sync with the appropriate filter in place for this - rather than inventing a whole new API for receiving data.  But it feels very ugly that the /sync call would block until the room_id in the filter was (implicitly) peeked.

Perhaps `/peek/snapshot` is good enough, assuming it returns the result in roughly the shape of a `/sync` response rather than inventing something new?

Reviewer: have added `/peek/snapshot` as a "future extension". We can add it if/when we have a good usecase for it.

>  3. [MSC1772](https://github.com/matrix-org/matrix-doc/pull/1772) (Groups as rooms) uses room joins to indicate your own membership, and peeks to query the group membership of other users.  Similarly to profiles, it's not clear that this should be per-user rather than per-device (and worse case, it's a matter of effectively opting in rather than trying to filter out peeks you don't care about).

Reviewer 11: We'd have to store somewhere which community rooms to peek... account data? That doesn't handle races well, though. But that's outside the scope of this MSC...

Author: mm, having a flag to say whether you want to peek on all devices could be quite a nice solution.

Reviewer 2: or for cases like this the server could implicitly assume you want to peek.

Reviewer 12: For microblogging style usecases, cross device peeks are what you want, and this has been one of the blockers for starting on that project (unclear if @Half-Shot is still interested, but I'd still like to see it happen). There's also other usecases: With groups-as-rooms peeking permanently across devices is also useful, because you can use publicly visible invite-only groups to filter your room list with them. Pretty sure there's more usecases for this still, so I think a flag like that should exist.

Reviewer: > or for cases like this the server could implicitly assume you want to peek.

based on what?

Reviewer: have added cross-device peeks as a "future extension".

# MSC2753: Peeking via Sync (Take 2)

## Problem

Peeking into rooms without joining them currently relies on the deprecated v1 /initialSync and /events APIs.

This poses the following issues:

 * Servers and clients must implement two separate sets of event-syncing logic, doubling complexity.
 * Peeking a room involves setting a stream of long-lived /events requests
   going. Having multiple streams is racey, competes for resources with the
   /sync stream, and doesn't scale given each room requires a new /events
   stream.
 * v1 APIs are deprecated and not implemented on new servers.

This MSC likely obsoletes [MSC1776](https://github.com/matrix-org/matrix-doc/pull/1776).

## Proposal

We add an CS API called `/peek/{roomIdOrAlias}`, very similar to `/join/{roomIdOrAlias}`.

Calling `/peek`:
 * Resolves the given room alias to a room ID, if needed.

>  * Adds the room (if permissions allow) to a new section of the /sync response called `peek` - but only for the device which called `/peek`.

Author: this is why i listed out the use cases to check below. in the majority of cases you only ever want to peek from a given device, not across a given account (unlike joins and most other actions).

 * The `peek` section is identical to `join`, but shows the live activity of rooms for which that device is peeking.

The API returns 404 on an unrecognised room ID or alias, or 403 if the room does not allow peeking.

Similar to `/join`, `/peek` lets you specify `server_name` querystring parameters to specify which server(s) to try to peek into the room via (when coupled with [MSC2444](https://github.com/matrix-org/matrix-doc/pull/2444)).

> If a user subsequently `/join`s the room they're peeking, we atomically move the room to the `join` block of the `/sync` response, allowing the client to build on the state and history it has already received without re-sending it down /sync.

Author: i'm not sure it's that difficult, given I just implemented it Dendrite (modulo debugging) without much pain. Sending state for newly joined rooms is already specialcased in syncapi, so i just added another clause to suppress sending it if you are on a device which is currently peeking that room.

> If a user subsequently `/join`s the room they're peeking, we atomically move the room to the `join` block of the `/sync` response, allowing the client to build on the state and history it has already received without re-sending it down /sync.

Author: the more i play with this in practice on dendrite, the more i think this is okay.

> To stop peeking, the user calls `/unpeek` on the room, similar to `/leave` or `/forget`.  This returns 200 on success, 404 on unrecognised ID, or 400 if the room was not being peeked in the first place.

Reviewer 13: What happens in the case where you're peeking and the room changes permissions to "no peeking allowed"?

Author: i guess you'd see the m.room.history_visibility or whatever event fly past and then receive nothing more. Perhaps we should autocancel the /peek too.

Reviewer 13: Clarification in this MSC would be fab!

Author: done

The act of joining a peeked room automatically cancels any ongoing /peeks by that user.

> Clients should check for any irrelevant peeked rooms on launch (left over from previous instances of the app) and explicitly `/unpeek` them to conserve resources.

Reviewer 13: I see you address this further down with using `/sync` as the heartbeat, which sounds good to me. The only problem I can see is that it will increase `/sync` latency if the server needs to re-peek before returning the response.

> Clients should check for any irrelevant peeked rooms on launch (left over from previous instances of the app) and explicitly `/unpeek` them to conserve resources.

Reviewer 10: It feels unclean to push this responsibility onto clients when we can already see if a client is calling `/sync` without `since`, or even if the client just hasn't synced for a while.

> Clients should check for any irrelevant peeked rooms on launch (left over from previous instances of the app) and explicitly `/unpeek` them to conserve resources.

Author: the /sync heartbeat is used to clean up s2s peeks - not c2s peeks (just because i close my laptop overnight doesn't mean my c2s peeks should all suddenly silently break). i like timo's idea of using an sinceless sync to clean up any /peeks, which should do the trick nicely here

> Clients should check for any irrelevant peeked rooms on launch (left over from previous instances of the app) and explicitly `/unpeek` them to conserve resources.

Reviewer 13: I dislike the idea of a sinceless sync cleaning up peeks because it's implicit - clients can sync without a since token for other reasons than clearing peek status.

> Clients should check for any irrelevant peeked rooms on launch (left over from previous instances of the app) and explicitly `/unpeek` them to conserve resources.

Author: hum. any better ideas? (i've added sinceless sync = cleanup to the MSC for now, but agreed it's a bit magical)

> Clients should check for any irrelevant peeked rooms on launch (left over from previous instances of the app) and explicitly `/unpeek` them to conserve resources.

Reviewer 13: Well you either need a client heartbeat or rely on the client to manually unpeek. Assuming we do a heartbeat, periodically hitting `/peek` seems reasonable to me - since after all if you want to be peeking for hours and hours maybe you should just join the room. Though I suppose the `join_rules: invite` stuff creates a weird "read-only" room a bit like having `events_default: 1`.

> Clients should check for any irrelevant peeked rooms on launch (left over from previous instances of the app) and explicitly `/unpeek` them to conserve resources.

Reviewer 14: I can certainly imagine uses for rooms where you want to peek forever, where you (as a room admin) don't want the member list to show users but you want them to read messages or get notified on events.

> Clients should check for any irrelevant peeked rooms on launch (left over from previous instances of the app) and explicitly `/unpeek` them to conserve resources.

Reviewer 4: Sinceless syncs are really *really* expensive right now? I'd really rather not encourage clients to do that until we have paginated sync.

> Clients should check for any irrelevant peeked rooms on launch (left over from previous instances of the app) and explicitly `/unpeek` them to conserve resources.

Reviewer: > < @Kegsay > clients can sync without a since token for other reasons than clearing peek status.

what are those reasons, other than "flush my cache", in which case it is reasonable to assume that they are also looking to clear their peek status.

> I'd really rather not encourage clients to do that until we have paginated sync.

I don't think we're encouraging clients to do sinceless syncs - just saying that if a client happens to do a sinceless sync then we will assume that they have forgotten which rooms they meant to be peeking into and clear the peek list for them.

(This would all fall out naturally if the "peek list" was sent as part of the `/sync` request as proposed at https://github.com/matrix-org/matrix-doc/pull/2753#discussion_r487133417)

> Clients should check for any irrelevant peeked rooms on launch (left over from previous instances of the app) and explicitly `/unpeek` them to conserve resources.

Reviewer: I feel like this thread is either resolved or a duplicate of https://github.com/matrix-org/matrix-doc/pull/2753#discussion_r480018048, so I'm going to resolve it.

The API returns 404 on an unrecognised room ID or alias, or 403 if the room does not allow peeking.  Rooms allow peeking if they have history_visibility of "world_readable".  N.B. join_rules do not affect peekability - it's possible to have an invite-only room which joe public can still peek into, if history_visibility has been set to "world_readable".

If the history_visibility of a room changes to not be world_readable, any peeks on the room are cancelled.

In order to clear up stale peeks if a client restarts without cleaning up nicely, the act of calling /sync without a `since` token must cancel any peeks created by that device.

Similar to `/join`, `/peek` lets you specify `server_name` querystring parameters to specify which server(s) to try to peek into the room via (when coupled with [MSC2444](https://github.com/matrix-org/matrix-doc/pull/2444)).

If a user subsequently `/join`s the room they're peeking, we atomically move the room to the `join` block of the `/sync` response, allowing the client to build on the state and history it has already received without re-sending it down /sync.

> To stop peeking, the user calls `/unpeek` on the room, similar to `/leave` or `/forget`.  This returns 200 on success, 404 on unrecognised ID, or 400 if the room was not being peeked in the first place.

Reviewer 2: `/forget` has no significant meaning when it comes to leaving a room anymore (since r0.5 iirc)

Reviewer: removed reference to /leave.

The act of joining a peeked room automatically cancels any ongoing /peeks by that user.

Clients should check for any irrelevant peeked rooms on launch (left over from previous instances of the app) and explicitly `/unpeek` them to conserve resources.

> 

Reviewer 2: Some boilerplate just to clarify the eventual spec PR:
```suggestion
The new `/peek` and `/unpeek` endpoints require authentication and can be ratelimited. Their 
responses are analogous to their `/join` and `/leave` counterparts (eg: `room_id` in the response 
and empty object when stopping).

```

Reviewer: added in a2551407

> Clients should check for any irrelevant peeked rooms on launch (left over from previous instances of the app) and explicitly `/unpeek` them to conserve resources.

Reviewer: > Clients should check for any irrelevant peeked rooms on launch

how should they do that? Is there a way for them to get a list of the currently-active peeks?

Author: the active peeks would be present in the /sync response. but i guess your point is that if the room is idle, it wouldn't show up at all? in which case, perhaps we need a way to enumerate the currently active ones, as you imply (or we mandate that empty peeked room sections are not elided from the /sync response)

Reviewer: this para is now removed, in favour of having peeks expire.
